### PR TITLE
New rule: avoid_annotating_with_dynamic

### PR DESCRIPTION
**Pull request by JPaulsen**
Fixes https://github.com/dart-lang/linter/issues/550

### CODE REVIEW

**Comment by JPaulsen** [comment]
@bwilkerson @pq I am having big problems with this rule, it is getting a lot of false positives when running in the SDK.

SDK lints: 185

Some false positives:

```dart


StringConversionSink startChunkedConversion(Sink<Object> sink) {
^^^^^^^^^^^^^^^^^^^^

Converter<List<int>, T> fuse<T>(Converter<String, T> next) {
                             ^

static Stream<ProcessSignal> _watchSignal(ProcessSignal signal) {
       ^^^^^^^^^^^^^^^^^^^^^
```

**Comment by bwilkerson** [comment]
Perhaps use textual comparison for return types as you do for parameter types. That might remove the false positives. But note the comment added to the issue.

**Comment by JPaulsen** [comment]
@bwilkerson nice!!!, 63 lints in the SDK, no false positives :), also I added some tests to have better test coverage.

**Comment by JPaulsen** [comment]
@bwilkerson I deleted the check of the return type, if you approve these changes, we should open a new issue to extend this rule in the future

File: lib/src/rules/avoid_annotating_with_dynamic_when_not_required.dart [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+}
+```
+
+''';
+
+class AvoidAnnotatingWithDynamicWhenNotRequired extends LintRule {
+  _Visitor _visitor;
+  AvoidAnnotatingWithDynamicWhenNotRequired()
+      : super(
+            name: 'avoid_annotating_with_dynamic_when_not_required',
```
Comment (bwilkerson) [inline lib/src/rules/avoid_annotating_with_dynamic_when_not_required.dart]: Perhaps shorten the name to "avoid_annotating_with_dynamic".

Comment (JPaulsen) [inline lib/src/rules/avoid_annotating_with_dynamic_when_not_required.dart]: Fixed

**Pull request merged**